ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
+// be summed to compute a section's offset.
+type asciiNgramOffset struct {
+	entries      []uint32 // (chara << 25 | charb << 18 | charc << 11 | length)
+	chunkOffsets []uint32 // offset for entries[i*asciiNgramOffsetChunkLength]
+}
+
+// asciiNgramOffsetChunkLength specifies how many entries share one initial offset.
+// It must be a power of 2, and was chosen empirically by measuring RAM usage:
+// 8: 4132MB, 16: 4047MB, 32: 4006MB, 64: 3992MB, 128: 3990MB
+const asciiNgramOffsetChunkLength = 32
+
+func makeAsciiNgramOffset(ngrams []uint32, offsets []uint32) *asciiNgramOffset {
+	// fmt.Println("makeAsciiNgramOffset", ngrams, offsets)
+	ao := &asciiNgramOffset{
+		entries:      make([]uint32, 0, len(ngrams)),
+		chunkOffsets: make([]uint32, 0, len(ngrams)/asciiNgramOffsetChunkLength),
+	}
+
+	for i, ng := range ngrams {
+		if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+			ao.chunkOffsets = append(ao.chunkOffsets, offsets[i])
+		}
+		length := offsets[i+1] - offsets[i]
+		for length > 0 {
+			if length >= ngramAsciiMaxSectionLength {
+				ao.entries = append(ao.entries, ng<<11|ngramAsciiMaxSectionLength)
+				length -= ngramAsciiMaxSectionLength
+			} else {
+				ao.entries = append(ao.entries, ng<<11|length)
+				length = 0
+			}
+			if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+				break
+			}
+		}
+	}
+
+	ao.entries = shrinkUint32Slice(ao.entries)
+	ao.chunkOffsets = shrinkUint32Slice(ao.chunkOffsets)
+
+	return ao
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
+func (a *asciiNgramOffset) Get(gram ngram) simpleSection {
+	if gram&ngramAsciiMask != gram {
+		return simpleSection{}
+	}
+	g := ngramAsciiToPacked(gram) << 11
+
+	idx := sort.Search(len(a.entries), func(i int) bool {
+		return a.entries[i] >= g
+	})
+
+	if idx == len(a.entries) || a.entries[idx]>>11 != g>>11 {
+		return simpleSection{}
+	}
+
+	length := a.entries[idx] & ngramAsciiMaxSectionLength
+	if length == ngramAsciiMaxSectionLength {
+		// this ascii ngram's section length is too large to be represented;
+		// repeate the Get() on the unicode map to get the correct result.
+		return simpleSection{
+			off: 0,
+			sz:  ngramAsciiMaxSectionLength,
+		}
+	}
+
+	chunkNum := idx / asciiNgramOffsetChunkLength
+	chunkBase := chunkNum * asciiNgramOffsetChunkLength
+	offset := a.chunkOffsets[chunkNum]
+	for i := chunkBase; i < idx; i++ {
+		offset += a.entries[i] & ngramAsciiMaxSectionLength
+	}
+
+	return simpleSection{
+		off: offset,
+		sz:  length,
+	}
+}
+
+func (a *asciiNgramOffset) DumpMap() map[ngram]simpleSection {
+	m := make(map[ngram]simpleSection, len(a.entries))
+	off := uint32(0)
+	for i, ent := range a.entries {
+		if i%asciiNgramOffsetChunkLength == 0 {
+			off = a.chunkOffsets[i/asciiNgramOffsetChunkLength]
+		}
+		length := ent & ngramAsciiMaxSectionLength
```

Comment:
you should be skipping lengths that are `ngramAsciiMaxSectionLength` to make this behave like calling Get to build the map. I realise in practice this doesn't matter since you call DumpMap on combined only, which will overwrite with the value from the unicode array. But still, I think it's more correct to not encode the large values into the map returned here.